stdlib/public/core/EmbeddedRuntime.swift

```diff
@@ -101,21 +101,31 @@ public func swift_isUniquelyReferenced_nonNull_native(object: UnsafeMutablePoint
 
 @_silgen_name("swift_retain")
 public func swift_retain(object: Builtin.RawPointer) -> Builtin.RawPointer {
+  return swift_retain_n(object: object, n: 1)
+}
+
+@_silgen_name("swift_retain_n")
+public func swift_retain_n(object: Builtin.RawPointer, n: UInt32) -> Builtin.RawPointer {
   if Int(Builtin.ptrtoint_Word(object)) == 0 { return object }
   let o = UnsafeMutablePointer<HeapObject>(object)
   // TODO/FIXME: Refcounting is not thread-safe, the following only works in single-threaded environments.
   if o.pointee.refcount == HeapObject.immortalRefCount { return o._rawValue }
-  o.pointee.refcount += 1
+  o.pointee.refcount += Int(n)
   return o._rawValue
 }
 
 @_silgen_name("swift_release")
 public func swift_release(object: Builtin.RawPointer) {
+  swift_release_n(object: object, n: 1)
+}
+
+@_silgen_name("swift_release_n")
+public func swift_release_n(object: Builtin.RawPointer, n: UInt32) {
   if Int(Builtin.ptrtoint_Word(object)) == 0 { return }
   let o = UnsafeMutablePointer<HeapObject>(object)
   // TODO/FIXME: Refcounting is not thread-safe, the following only works in single-threaded environments.
   if o.pointee.refcount == HeapObject.immortalRefCount { return }
-  o.pointee.refcount -= 1
+  o.pointee.refcount -= Int(n)
   if (o.pointee.refcount & HeapObject.refcountMask) == 0 {
```

Comment:
I think it would make sense to add a check for a resulting negative ref count, which should never happen. This could catch compiler bugs and security attacks.
Also, it shouldn't be required to mask the refcount field anymore, because the immortal case is already checked before.

```
  if o.pointee.refcount <= 0 {
    if o.pointee.refcount < 0 {
      fatalError("negative refcount")
    }
    _swift_embedded_invoke_heap_object_destroy(o)
  }
```

Reply:
Added the underflow check, but I think I still need the masking -- it's for the "doNotFree" bit, if it's set, then the refcount integer value actually does show up as negative unless I mask it. Or am I missing something?

Reply:
Ah, you are right! I missed that we have the doNotFree bit